contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,94 @@
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+Reducing the code complexity and improving the correctness will also make it simpler to move this logic
+to the server as part of an API in the future, while also addressing the immediate needs of the client.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
+  end state that will result from apply, and implementing a diff operation has been exceedingly
+  copmlex.
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+- Patches provide weaker optimistic locking guarantees than PUT.  When encountering an optimistic
+  lock failure in apply, it is easy to re-read the object and try to update it again.  If the
+  apply command tries to modify a field that was modified since apply read the object, it will fail
+  and optionally be able to show this to a user, whereas this conflict may not be detect with a PATCH
+  operation.
+
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
+
+#### Trade offs
+
+Good:
+
+- Simpler code to maintain
+- Behaves correctly in more instances
+- Easy to look at the output and determine the final state
+
+Bad:
+
+- Larger request sizes to the server than when sending a patch
```

Comment:
We had also discussed sending the full configured resource decorated with deletion directives, which would have been slightly larger.